subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/asm/ClassDependenciesVisitor.java

```diff
@@ -35,33 +38,98 @@
+    }
+
+    public ClassDependenciesVisitor(Set<Integer> constantsCollector, Set<Integer> literalsCollector, Set<String> types, Predicate<String> typeFilter, ClassReader reader) {
         super(API);
         this.constants = constantsCollector;
         this.literals = literalsCollector;
-        this.superTypes = Sets.newHashSet();
+        this.types = types;
+        this.superTypes = types == null ? null : Sets.<String>newHashSet();
         this.annotationVisitor = literals == null ? null : new LiteralRecordingAnnotationVisitor();
         this.literalAdapter = literals == null ? null : new LiteralAdapter();
+        this.typeFilter = typeFilter;
+        if (reader != null) {
+            collectClassDependencies(reader);
+        }
     }
 
     @Override
     public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
         isAnnotationType = isAnnotationType(interfaces);
-        superTypes.add(Type.getObjectType(superName).getClassName());
+        String type = slashyTypeOf(superName);
+        maybeAddSuperType(type);
+        maybeAddDependentType(type);
         for (String s : interfaces) {
-            superTypes.add(Type.getObjectType(s).getClassName());
+            String interfaceType = slashyTypeOf(s);
+            maybeAddDependentType(interfaceType);
+            maybeAddSuperType(interfaceType);
+        }
+
+    }
+
+    // performs a fast analysis of classes referenced in bytecode (method bodies)
+    // allowing us to implement a costly visitor and potentially missing edge cases
+    private void collectClassDependencies(ClassReader reader) {
+        char[] charBuffer = new char[reader.getMaxStringLength()];
+        for (int i = 1; i < reader.getItemCount(); i++) {
+            int itemOffset = reader.getItem(i);
+            if (itemOffset > 0 && reader.readByte(itemOffset - 1) == 7) {
+                // A CONSTANT_Class entry, read the class descriptor
+                String classDescriptor = reader.readUTF8(itemOffset, charBuffer);
+                Type type = Type.getObjectType(classDescriptor);
+                while (type.getSort() == Type.ARRAY) {
+                    type = type.getElementType();
+                }
+                if (type.getSort() != Type.OBJECT) {
+                    // A primitive type
+                    continue;
+                }
+                String name = type.getClassName();
+                maybeAddDependentType(name);
+            }
+        }
+    }
+
+    protected void maybeAddSuperType(String type) {
+        if (superTypes != null && typeFilter.apply(type)) {
+            superTypes.add(type);
         }
     }
 
+    protected void maybeAddDependentType(String type) {
+        if (types != null && typeFilter.apply(type) && !PRIMITIVES.contains(type)) {
+            types.add(type);
+        }
+    }
+
+    protected String slashyTypeOf(String slashyStyleDesc) {
```

Comment:
This method reads like we are creating a "slashy type", but actually we're doing the opposite: Converting from slashes to class name